Child device firmware management specification #1998 (#2007)
thin-edge/thin-edge.io: docs/src/references/c8y-firmware-management.md

# Device Firmware Management using Cumulocity

Thin-edge provides an operation plugin to
[manage device firmware using Cumulocity](https://cumulocity.com/guides/users-guide/device-management/#firmware-repo).
Firmware management is currently supported only for child devices and not for the main tedge device.

## Installation

> The plugin will be installed at `/usr/bin/c8y-firmware-plugin` by the debian package.
> The systemd service definition files for the plugin are also installed at `/lib/systemd/system/c8y-firmware-plugin.service`.

Reviewer: Do we need mention the exact paths here? Or would it just be enough to say:

```suggestion
The plugin includes a single executable called `c8y-firmware-plugin` and a systemd service definition with the same name.
```

Author: If it was user facing how-to guide, I'd have definitely gone with your version. But, since this is a plugin specification which the implementor should have used to create the debian package definition as well (if this doc existed before the dev implemented it  ;), of course), the details should be specific, right?

Reviewer: Fair enough. You can leave it as is then :)

No operations files are created under `/etc/tedge/operations/c8y/`
as this plugin doesn't support firmware updates for the tedge device.
Operation files for child devices must be created as part of their bootstrap process, which is explained later.

## Configuration

The plugin supports a single tedge configuration named `firmware.child.update.timeout`,
that defines the amount of time the plugin wait for a child device to finish a firmware update once the request is delivered.

> The default timeout value is `3600s` and can be updated with:

Reviewer: Does this mean the tedge config command would accept the value `3600s`, or is the unit of the config just in seconds, but only the value of `3600` would be accepted. For example:

```
sudo tedge config set firmware.child.update.timeout 3600s
```

If `3600s` is not accepted as a valid value then it would be better to be explicit about it.

```suggestion
The default timeout (in seconds) value is `3600` and can be updated with:
```

```shell

> sudo tedge config set firmware.child.update.timeout <another-value>

Reviewer: Dependent on the previous comment about whether the time units (`s`) can be part of the value or not. If not then having a hint for the user that it is in seconds would be useful:

```suggestion
sudo tedge config set firmware.child.update.timeout <timeout_seconds>
```

  * to handle network failures during the download even on flaky networks
  * to publish the downloaded firmware files over a local HTTP server and make them available to the child-devices,
  * to notify the child-devices when firmware updates are available,
  * to receive forward the firmware update status updates from the child devices to the cloud
* By contrast, the plugin is not responsible for:
  * checking the integrity of the downloaded file which is a third-party binary
  * installing the firmware files on the child-devices.
* For each child-device, a device-specific software component, referred to as a `child-device-connector`, 
  is required to listen for firmware update related MQTT notifications from the plugin
  and behave accordingly along the protocol defined by this plugin.
  * Being specific to each type of child devices, this software has to be implemented specifically.
  * This software can be installed on the child device.
  * This software can also be installed on the main device,
    when the target device cannot be altered or connected to the main device over MQTT and HTTP.

### Child device connector connecting to thin-edge device

> The `child-device-connector` interacts with thin-edge over its MQTT and HTTP APIs.

Reviewer: A sentence about what the `child-device-connector` is would be useful to give it more context in case if user's didn't see the first reference to it in the list of responsibilities.

So maybe add something like this as the first sentence of the paragraph.

```markdown
The `child-device-connector` has the responsibility of handling firmware requests sent by the thin-edge and translates it to the relevant 3rd-party API to apply the contents of the firmware update request.
```

In cases where the child device connector is installed alongside thin-edge on the same device,
these APIs can be accessed via a local IP or even `127.0.0.1`.
The MQTT APIs are exposed via port 1883 and the HTTP APIs are exposed via port 8000, by default.
When the child device connector is running directly on the external child device,
the MQTT and HTTP APIs of thin-edge need to be accessed over the network using its IP address and ports,
which are configured using the tedge config settings `mqtt.client.host` or `mqtt.client.port` for MQTT
and `http.address` and `http.port` for HTTP.

### Child devices declaring firmware management support

> But for child devices, the operation files must be created under `/etc/tedge/operations/c8y/$CHILD_DEVICE_ID`.

Reviewer: Just to make sure users don't interpret the `$` as a literal char (it happens).

```suggestion
But for child devices, the operation files must be created under `/etc/tedge/operations/c8y/$CHILD_DEVICE_ID`, where `$CHILD_DEVICE_ID` should be replaced with the child's identity.
```

These files are not created by the plugin itself, but must be created by the child device connector or
by any other means for each child device as follows:

```shell
$ tree /etc/tedge/operations/c8y
/etc/tedge/operations/c8y
|-- child-1

> |   `-- c8y_Firmware

Reviewer: ```suggestion
|   └-- c8y_Firmware
```

> |-- child-1
> |   `-- c8y_Firmware
> `-- child-2
>     `-- c8y_Firmware

Reviewer: Just some cosmetic improvements to the tree.

```suggestion
|-- child-1
|   └-- c8y_Firmware
└-- child-2
    └-- c8y_Firmware
```

```

Using the file watcher mechanism, the c8-mapper picks up the creation of these operation files
and report them as supported operations for the child device as follows:

```shell
$ tedge mqtt sub 'c8y/s/us/#'
[c8y/s/us/child-1] 114,c8y_Firmware
[c8y/s/us/child-2] 114,c8y_Firmware

> ```

Reviewer: Can we be more generic here and just say that the c8y mapper detects filesystem changes and will register the devices in Cumulocity accordingly (Is the implementation details about sending smart rest message really needed?).

Anyway we should use `c8y mapper` instead of `c8y-mapper` (after fixing the typo), so that we don't confuse users thinking that there is a `c8y-mapper` binary. And we don't need to wrap it in backticks as that implies an official name.

Author: Agree. Since sending the Smartrest messages is the responsibility of the mapper and not this plugin, those details are not needed here.

    participant Tedge Agent
    participant Child Device Connector
    

        C8Y Cloud->>C8Y Firmware Plugin: MQTT: c8y_Firmware request with `child-id` and `c8y.url` for the updated firmware file
        C8Y Firmware Plugin->>C8Y Cloud: Download the firmware file from `c8y.url` to firmware cache
        C8Y Firmware Plugin->>Tedge Agent: Symlink the cached firmware file to file-transfer repository and generate a `tedge.url` for it
        C8Y Firmware Plugin->>Child Device Connector: MQTT: firmware_update request to `child-id` with `tedge.url`

        Child Device Connector ->> C8Y Firmware Plugin: MQTT: firmware_update response with operation status: "executing" 
        C8Y Firmware Plugin ->> C8Y Cloud: MQTT: Update c8y_Firmware operation status to "EXECUTING"
        Child Device Connector->>Tedge Agent: HTTP: Download the firmware file from `tedge.url`
        Child Device Connector->>Child Device Connector: Apply downloaded firmware file
        Child Device Connector ->> C8Y Firmware Plugin: MQTT: firmware_update response with operation status: "successful" 

        C8Y Firmware Plugin ->> C8Y Cloud: MQTT: Update c8y_Firmware operation status to "SUCCESSFUL"

>         C8Y Firmware Plugin ->> C8Y Firmware Plugin: Remove the symlink from file-transfer repository but keep the cached firmware copy for reuse

Reviewer 2: I don't think it removes symlink... instead deleting a firmware info file under `FIRMWARE_OP_PATH`.

Author: We should be removing those symlinks as it makes it easier for the device admin to cleanup the firmware cache, when needed by just looking at the files that don't have any symlinks pointed to it.

1. The plugin, on recept of a `c8y_Firmware` request from Cumulocity for a child device named `$CHILD_DEVICE_ID`
   in the format `515,$CHILD_DEVICE_ID,$FIRMWARE_NAME,$FIRMWARE_VERSION,$FIRMWARE_URL`
   1. Validate if the same firmware update operation is already in progress
      by iterating over all the operation files in the `$FIRMWARE_OP_PATH` directory.
      The operation files contains the last `firmware_update` request's JSON payload along with the `device` ID.
      If an operation file with the `device` id, `name`, `version` and `url` fields matching
      the incoming `$FIRMWARE_NAME`,`$FIRMWARE_VERSION` and `$FIRMWARE_URL` is found,
      the same request is re-sent to the child device by just incrementing the `attempt` count value.
      The operation file content is also overwritten the with updated `attempt` count. 
   1. If a pending operation match is not found, do a look up if the firmware file for the given url already exists
      in its firmware cache at `$FIRMWARE_CACHE_PATH`.
      The file name for the lookup is derived from the SHA-256 digest of the firmware url.
      If a cached copy exists, the download is skipped.

>    1. If not found in the firmware cache, the plugin downloads the firmware file from the `url` to `$TEDGE_TMP_PATH`.

Reviewer 2: After we changed `Downloader` implementation, the plugin (to be precise downloader) downloads the file to the final destination directory (so here `$TEDGE_DATA_PATH/cache`) with the name `<name>.tmp`. 

Here's downloader code: https://github.com/thin-edge/thin-edge.io/blob/main/crates/common/download/src/download.rs#L302

Actually we should remove loading `tmp` path from firmware-plugin...

   1. On successful download, the file is moved to the firmware cache at `$FIRMWARE_CACHE_PATH`
      with the name derived from the SHA-256 digest of the firmware url.
1. The cached firmware file is published via the file-transfer repository of `tedge-agent` 
   by creating a symlink to the cached firmware file is created in the file-transfer repository at
   `$FILE_TRANSFER_REPO/$CHILD_DEVICE_ID/firmware_update/$FILE_ID` making this file available via
   the HTTP endpoint: `http://$TEDGE_HTTP_ADDRESS/tedge/file-transfer/$CHILD_DEVICE_ID/firmware_update/$FILE_ID`.
1. Once the updated firmware file is published via the HTTP file transfer service,
   the plugin send the `firmware_update` request to the child device connector by publishing an MQTT message:
   * Topic: `tedge/$CHILD_DEVICE_ID/commands/req/firmware_update`
   * The payload is a JSON record with the following fields
     * `id`: A unique id generated by the plugin
     * `name`: Name of the firmware received in the cloud request

>      * `version`:

Reviewer 2: Describe something about version :D

     * `url`: The file-transfer service entry URL(`http://$TEDGE_HTTP_ADDRESS/tedge/file-transfer/$CHILD_DEVICE_ID/firmware_update/$FILE_ID`)
     * `sha256`: The SHA-256 checksum of the firmware file served via the `url`

>      * `attempt`: The count that indicates if this request is being resent or not

Reviewer 2: Maybe worth mentioning that the counts start from `1`? (I mean not 0)

      * Topic: `tedge/$CHILD_DEVICE_ID/commands/res/firmware_update`
      * Payload must be a JSON record with the following fields
        * `id`: The `id` of the request
        * `status`: "executing"
   1. `GET`s the firmware file from the `url` specified by the notification message.
   1. Validate the integrity of the downloaded binary by matching its SHA-256 hash value
      against the `sha256` checksum value received in the request.
   1. Apply the downloaded firmware file update on the device using whatever device specific protocol.
1. After applying the update, send the final operation status update to thin-edge via MQTT:
   1. Topic: `tedge/$CHILD_DEVICE_ID/commands/res/firmware_update`
   1. The payload must be a JSON record with the following fields:
      * `id`: The `id` of the request received
      * `status`: `successful` or `failed` based on the result of updating the firmware
      * `reason`: The reason for the failure, applicable only for `failed` status.
1. On reception of an operation status message, the plugin maps it to SmartREST and forwards it to the cloud.

>    * When a `success` or `failed` status message is finally received,

Reviewer 2: ```suggestion
   * When a `successful` or `failed` status message is finally received,
```

>      then the plugin cleans up the corresponding operation file at `$FIRMWARE_OP_PATH/$OP_ID` and

Reviewer 2: As I said above, the current plugin implementation doesn't clean up symlink. If we need it, we need to change the code.

Author: I think we need to change the code.

Reviewer 3: So a ticket needs to be created when this PR is merged.

Author: Done: https://github.com/thin-edge/thin-edge.io/issues/2023

     the firmware file entry in the file transfer repository at `$FILE_TRANSFER_REPO/$CHILD_DEVICE_ID/firmware_update/$FILE_ID`.
   * If a notification message is received while none is expected,

>      i.e with an operation `id` that doesn't exist  `TEDGE_HTTP_ADDRESS_ROOT/$CHILD_DEVICE_ID/firmware_update/`,

Reviewer 2: I think it's the mistake on the path?
```suggestion
     i.e with an operation `id` that doesn't exist  `$TEDGE_DATA_PATH/firmware/<id>`,
```

     then this notification message is ignored.      

## Logging

The plugin logs its progress and errors on to its `stderr`.

* All upload and download operation requests are logged, when received and when completed,
  with one line per file.
* All changes to the list of managed file is logged, one line per change.

> * All errors are reported with the operation context (upload or download? which file?).

Reviewer: What does the last line mean exactly, it seems incomplete.

Author: Yeah, that's the unfortunate side-effect of context switching in the middle of editing copy-pasted content ;)

# Device Firmware Management using Cumulocity

Thin-edge provides an operation plugin to
[manage device firmware using Cumulocity](https://cumulocity.com/guides/users-guide/device-management/#firmware-repo).
Firmware management is currently supported only for child devices and not for the main tedge device.

- The firmware update operations are defined and triggered from Cumulocity
- Thin-edge acts as the proxy between Cumulocity and the child device
  facilitating the routing of firmware update requests as well as the transfer of firmware binary files from cloud to the device.
- Since thin-edge can not install the firmware on the child device directly,
  as that is a device specific action that may require additional interactions using device specific protocols as well,
  an additional piece of software, referred to as `child-device-connector` in the rest fo this doc,

>   must be developed by the child device admin to perform the actual installation itself in coordination with thin-edge.

Author: Reversed the flow, as suggested.

   1. Apply the downloaded firmware file update on the device using whatever device specific protocol.
1. After applying the update, send the final operation status update to thin-edge via MQTT:
   1. Topic: `tedge/$CHILD_DEVICE_ID/commands/res/firmware_update`
   1. The payload must be a JSON record with the following fields:
      * `id`: The `id` of the request received
      * `status`: `successful` or `failed` based on the result of updating the firmware
      * `reason`: The reason for the failure, applicable only for `failed` status.
1. On reception of an operation status message, the plugin maps it to SmartREST and forwards it to the cloud.
   * When a `successful` or `failed` status message is finally received,
     then the plugin cleans up the corresponding operation file at `$FIRMWARE_OP_PATH/$OP_ID` and
     the firmware file entry in the file transfer repository at `$FILE_TRANSFER_REPO/$CHILD_DEVICE_ID/firmware_update/$FILE_ID`.
   * If a notification message is received while none is expected,
     i.e with an operation `id` that doesn't exist at `$TEDGE_DATA_PATH/firmware/<id>`,
     then this notification message is deemed stale and ignored.

## Logging

> The plugin logs its progress and errors on to its `stderr`.

Reviewer 4: I want clarification, the progress `log messages` are just printed onto the `stderr` and not logged into any files right?

Author: Not to any file. When running as a `systemd` service, the logs can be retrieved using `journalctl`.
